stream_profile.go

```diff
@@ -0,0 +1,179 @@
+package iocontrol
+
+import (
+	"io"
+	"sync/atomic"
+	"time"
+
+	"github.com/benbjohnson/clock"
+)
+
+// TimeProfile contains information about the timings involved in
+// the exchange of bytes between an io.Writer and io.Reader.
+type TimeProfile struct {
+	WaitRead  time.Duration
+	WaitWrite time.Duration
+	Total     time.Duration
+}
+
+// Profile will wrap a writer and reader pair and profile where
+// time is spent: writing or reading. The result is returned when
+// the `done` func is called. The `done` func can be called multiple
+// times.
+//
+// There is a small performance overhead of ~µs per Read/Write call.
+// This is negligible in most I/O workload. If the overhead is too
+// much for your needs, use the `ProfileSample` call.
+func Profile(w io.Writer, r io.Reader) (pw io.Writer, pr io.Reader, done func() TimeProfile) {
+	return profile(clock.New(), w, r)
+}
+
+func profile(clk clock.Clock, w io.Writer, r io.Reader) (io.Writer, io.Reader, func() TimeProfile) {
+
+	preciseWriter := &preciseTimedWriter{clk: clk, w: w}
+	preciseReader := &preciseTimedReader{clk: clk, r: r}
+
+	start := clk.Now()
+	return preciseWriter, preciseReader, func() TimeProfile {
+		return TimeProfile{
+			Total:     clk.Now().Sub(start),
+			WaitRead:  preciseReader.WaitRead(),
+			WaitWrite: preciseWriter.WaitWrite(),
+		}
+	}
+}
+
+type preciseTimedReader struct {
+	clk   clock.Clock
+	r     io.Reader
+	sumNS int64
+}
+
+func (t *preciseTimedReader) WaitRead() time.Duration {
+	return time.Duration(atomic.LoadInt64(&t.sumNS))
+}
+
+func (t *preciseTimedReader) Read(p []byte) (int, error) {
+	start := t.clk.Now()
+	n, err := t.r.Read(p)
+	atomic.AddInt64(&t.sumNS, t.clk.Now().Sub(start).Nanoseconds())
+	return n, err
+}
+
+type preciseTimedWriter struct {
+	clk   clock.Clock
+	w     io.Writer
+	sumNS int64
+}
+
+func (t *preciseTimedWriter) WaitWrite() time.Duration {
+	return time.Duration(atomic.LoadInt64(&t.sumNS))
+}
+
+func (t *preciseTimedWriter) Write(p []byte) (int, error) {
+	start := t.clk.Now()
+	n, err := t.w.Write(p)
+	atomic.AddInt64(&t.sumNS, t.clk.Now().Sub(start).Nanoseconds())
+	return n, err
+}
+
+// sampling, high performance profiler
+
+const (
+	state_runtime uint32 = iota
+	state_blocked
+)
+
+// ProfileSample will wrap a writer and reader pair and collect
+// samples of where time is spent: writing or reading. The result
+// is an approximation that is returned when the `done` func is
+// called. The `done` func can be called *only once*.
```

Comment:
Might be worth making the `done` func panic if called twice, just to reinforce this comment.